pytissueoptics/rayscattering/opencl/exampleSimple.py

```diff
@@ -0,0 +1,25 @@
+import numpy as np
+
+from pytissueoptics import *
```

Comment:
This does not run for me.

```
Traceback (most recent call last):
  File "/Users/dccote/GitHub/PyTissueOptics/pytissueoptics/rayscattering/opencl/CLConfig.py", line 128, in _autoSetNWorkUnits
    optimalNWorkUnits = computeOptimalNWorkUnits()
  File "/Users/dccote/GitHub/PyTissueOptics/pytissueoptics/rayscattering/opencl/testWorkUnits.py", line 46, in computeOptimalNWorkUnits
    source.propagate(scene, logger=logger, showProgress=False)
  File "/Users/dccote/GitHub/PyTissueOptics/pytissueoptics/rayscattering/source.py", line 35, in propagate
    IPP = self._getAverageInteractionsPerPhoton(scene)
  File "/Users/dccote/GitHub/PyTissueOptics/pytissueoptics/rayscattering/source.py", line 62, in _getAverageInteractionsPerPhoton
    self._measureIPP(scene)
  File "/Users/dccote/GitHub/PyTissueOptics/pytissueoptics/rayscattering/source.py", line 79, in _measureIPP
    self._propagateOpenCL(estimatedIPP, scene, tempLogger, showProgress=False)
  File "/Users/dccote/GitHub/PyTissueOptics/pytissueoptics/rayscattering/source.py", line 99, in _propagateOpenCL
    self._photons.propagate(IPP=IPP, verbose=showProgress)
  File "/Users/dccote/GitHub/PyTissueOptics/pytissueoptics/rayscattering/opencl/CLPhotons.py", line 60, in propagate
    program.launchKernel(kernelName="propagate", N=np.int32(params.workItemAmount),
  File "/Users/dccote/GitHub/PyTissueOptics/pytissueoptics/rayscattering/opencl/CLProgram.py", line 30, in launchKernel
    self._build(CLObjects)
  File "/Users/dccote/GitHub/PyTissueOptics/pytissueoptics/rayscattering/opencl/CLProgram.py", line 59, in _build
    self._program = cl.Program(self._context, sourceCode).build()
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/pyopencl/__init__.py", line 539, in build
    self._prg, was_cached = self._build_and_catch_errors(
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/pyopencl/__init__.py", line 587, in _build_and_catch_errors
    raise err
pyopencl._cl.RuntimeError: clBuildProgram failed: BUILD_PROGRAM_FAILURE - clBuildProgram failed: BUILD_PROGRAM_FAILURE - clBuildProgram failed: BUILD_PROGRAM_FAILURE

Build on <pyopencl.Device 'Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz' on 'Apple' at 0xffffffff>:

<program source>:61:6: warning: no previous prototype for function 'wangHash'
uint wangHash(uint seed){
     ^
<program source>:70:7: warning: no previous prototype for function 'getRandomFloatValue'
float getRandomFloatValue(__global unsigned int *seeds, unsigned int id){
      ^
<program source>:85:6: warning: no previous prototype for function 'normalizeVectorLocal'
void normalizeVectorLocal(float3 *vector){
     ^
<program source>:94:6: warning: no previous prototype for function 'normalizeVectorGlobal'
void normalizeVectorGlobal(__global float3 *vector){
     ^
<program source>:103:6: warning: no previous prototype for function 'rotateAroundAxisGlobal'
void rotateAroundAxisGlobal(__global float3 *mainVector, __global float3 *axisVector, float theta){
     ^
<program source>:128:6: warning: no previous prototype for function 'rotateAroundAxisLocal'
void rotateAroundAxisLocal(float3 *mainVector, float3 *axisVector, float theta){
     ^
<program source>:153:6: warning: no previous prototype for function 'rotateAround'
void rotateAround(float3 *mainVector, float3 *axisVector, float theta){
     ^
<program source>:178:8: warning: no previous prototype for function 'getAnyOrthogonalGlobal'
float3 getAnyOrthogonalGlobal(__global float3 *vector){
       ^
<program source>:185:8: warning: no previous prototype for function 'getAnyOrthogonal'
float3 getAnyOrthogonal(float3 *vector){
       ^
<program source>:210:7: warning: no previous prototype for function 'getScatteringDistance'
float getScatteringDistance(float mu_t, float randomNumber){
      ^
<program source>:214:7: warning: no previous prototype for function 'getScatteringAnglePhi'
float getScatteringAnglePhi(float randomNumber){
      ^
<program source>:219:7: warning: no previous prototype for function 'getScatteringAngleTheta'
float getScatteringAngleTheta(float g, float randomNumber){
      ^
<program source>:229:18: warning: no previous prototype for function 'getScatteringAngles'
ScatteringAngles getScatteringAngles(float rndPhi, float rndTheta,__global Photon *photons,
                 ^
<program source>:255:13: error: global variables must have a constant address space qualifier
const float EPS = 0.00001f;
            ^
<program source>:256:13: error: global variables must have a constant address space qualifier
const float EPS_CORRECTION = 0.0005f;
            ^
<program source>:257:13: error: global variables must have a constant address space qualifier
const float EPS_PARALLEL = 0.00001f;
            ^
<program source>:258:13: error: global variables must have a constant address space qualifier
const float EPS_SIDE = 0.000001f;
            ^
<program source>:300:21: warning: no previous prototype for function '_getBBoxIntersection'
GemsBoxIntersection _getBBoxIntersection(Ray ray, float3 minCornerVector, float3 maxCornerVector) {
                    ^
<program source>:370:6: warning: no previous prototype for function '_findBBoxIntersectingSolids'
void _findBBoxIntersectingSolids(Ray ray, Scene *scene, uint gid){
     ^
<program source>:387:6: warning: no previous prototype for function '_sortSolidCandidates'
void _sortSolidCandidates(Scene *scene, uint gid) {
     ^
<program source>:412:10: warning: no previous prototype for function '_getTriangleIntersection'
HitPoint _getTriangleIntersection(Ray ray, float3 v1, float3 v2, float3 v3) {
         ^
<program source>:460:14: warning: no previous prototype for function '_findClosestPolygonIntersection'
Intersection _findClosestPolygonIntersection(Ray ray, uint solidID,
             ^
<program source>:489:7: warning: no previous prototype for function '_cotangent'
float _cotangent(float3 v0, float3 v1, float3 v2) {
      ^
<program source>:495:6: warning: no previous prototype for function 'setSmoothNormal'
void setSmoothNormal(Intersection *intersection, __global Triangle *triangles, __global Vertex *vertices) {
     ^
<program source>:517:6: warning: no previous prototype for function '_composeIntersection'
void _composeIntersection(Intersection *intersection, Ray *ray, Scene *scene) {
     ^
<program source>:528:14: warning: no previous prototype for function 'findIntersection'
Intersection findIntersection(Ray ray, Scene *scene, uint gid) {
             ^
<program source>:586:7: warning: no previous prototype for function '_getReflectionCoefficient'
float _getReflectionCoefficient(float n1, float n2, float thetaIn) {
      ^
<program source>:613:6: warning: no previous prototype for function '_getIsReflected'
bool _getIsReflected(float nIn, float nOut, float thetaIn, __global uint *seeds, uint gid) {
     ^
<program source>:622:7: warning: no previous prototype for function '_getReflectionDeflection'
float _getReflectionDeflection(float thetaIn) {
      ^
<program source>:626:7: warning: no previous prototype for function '_getRefractionDeflection'
float _getRefractionDeflection(float nIn, float nOut, float thetaIn) {
      ^
<program source>:632:6: warning: no previous prototype for function '_createFresnelIntersection'
void _createFresnelIntersection(FresnelIntersection* fresnelIntersection,
     ^
<program source>:643:21: warning: no previous prototype for function 'computeFresnelIntersection'
FresnelIntersection computeFresnelIntersection(float3 rayDirection, Intersection *intersection,
                    ^
<program source>:677:11: error: global variables must have a constant address space qualifier
const int NO_SOLID_ID = -1;
          ^
<program source>:678:11: error: global variables must have a constant address space qualifier
const int NO_SURFACE_ID = -1;
          ^
<program source>:680:6: warning: no previous prototype for function 'moveBy'
void moveBy(float distance, __global Photon *photons, uint photonID){
     ^
<program source>:684:6: warning: no previous prototype for function 'scatterBy'
void scatterBy(float phi, float theta, __global Photon *photons, uint photonID){
     ^
<program source>:689:6: warning: no previous prototype for function 'decreaseWeightBy'
void decreaseWeightBy(float delta_weight, __global Photon *photons, uint photonID){
     ^
<program source>:693:6: warning: no previous prototype for function 'interact'
void interact(__global Photon *photons, __constant Material *materials, __global DataPoint *logger,
     ^
<program source>:705:6: warning: no previous prototype for function 'scatter'
void scatter(__global Photon *photons, __constant Material *materials, __global uint *seeds, __global DataPoint *logger,
     ^
<program source>:717:6: warning: no previous prototype for function 'roulette'
void roulette(float weightThreshold, __global Photon *photons, __global uint *seeds, uint gid, uint photonID){
     ^
<program source>:730:6: warning: no previous prototype for function 'reflect'
void reflect(FresnelIntersection *fresnelIntersection, __global Photon *photons, uint photonID){
     ^
<program source>:731:18: error: passing '__global __float3 *' to parameter of type '__float3 *' changes address space of pointer
    rotateAround(&photons[photonID].direction, &fresnelIntersection->incidencePlane, fresnelIntersection->angleDeflection);
                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
<program source>:153:27: note: passing argument to parameter 'mainVector' here
void rotateAround(float3 *mainVector, float3 *axisVector, float theta){
                          ^
<program source>:734:6: warning: no previous prototype for function 'refract'
void refract(FresnelIntersection *fresnelIntersection, __global Photon *photons, uint photonID){
     ^
<program source>:735:18: error: passing '__global __float3 *' to parameter of type '__float3 *' changes address space of pointer
    rotateAround(&photons[photonID].direction, &fresnelIntersection->incidencePlane, fresnelIntersection->angleDeflection);
                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
<program source>:153:27: note: passing argument to parameter 'mainVector' here
void rotateAround(float3 *mainVector, float3 *axisVector, float theta){
                          ^
<program source>:738:6: warning: no previous prototype for function 'logIntersection'
void logIntersection(Intersection *intersection, __global Photon *photons, __global Surface *surfaces,
     ^
<program source>:766:7: warning: no previous prototype for function 'reflectOrRefract'
float reflectOrRefract(Intersection *intersection, __global Photon *photons, __constant Material *materials,
      ^
<program source>:772:30: warning: comparison of integers of different signs: 'int' and '__global uint' (aka '__global unsigned int')
    if (solidIDTowardsNormal != photons[photonID].solidID) {
        ~~~~~~~~~~~~~~~~~~~~ ^  ~~~~~~~~~~~~~~~~~~~~~~~~~
<program source>:810:7: warning: no previous prototype for function 'propagateStep'
float propagateStep(float distance, __global Photon *photons, __constant Material *materials, Scene *scene,
      ^
<program source>:838:38: warning: comparison of integers of different signs: 'int' and '__global uint' (aka '__global unsigned int')
            if (solidIDTowardsNormal != photons[photonID].solidID) {
                ~~~~~~~~~~~~~~~~~~~~ ^  ~~~~~~~~~~~~~~~~~~~~~~~~~

(options: -I /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/pyopencl/cl)
(source saved as /var/folders/vw/xyjt8vln3yb962l7lzwbgvrc0000gn/T/tmpj4dgkn0z.cl)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/dccote/GitHub/PyTissueOptics/pytissueoptics/rayscattering/opencl/exampleSimple.py", line 14, in <module>
    source = DivergentSource(position=Vector(0, 0, -0.1), direction=Vector(0, 0, 1), N=N,
  File "/Users/dccote/GitHub/PyTissueOptics/pytissueoptics/rayscattering/source.py", line 227, in __init__
    super().__init__(position=position, direction=direction, diameter=diameter, N=N,
  File "/Users/dccote/GitHub/PyTissueOptics/pytissueoptics/rayscattering/source.py", line 169, in __init__
    super().__init__(position=position, N=N, useHardwareAcceleration=useHardwareAcceleration)
  File "/Users/dccote/GitHub/PyTissueOptics/pytissueoptics/rayscattering/source.py", line 26, in __init__
    useHardwareAcceleration = validateOpenCL()
  File "/Users/dccote/GitHub/PyTissueOptics/pytissueoptics/rayscattering/opencl/__init__.py", line 15, in validateOpenCL
    CONFIG.validate()
  File "/Users/dccote/GitHub/PyTissueOptics/pytissueoptics/rayscattering/opencl/CLConfig.py", line 58, in validate
    self._autoSetNWorkUnits()
  File "/Users/dccote/GitHub/PyTissueOptics/pytissueoptics/rayscattering/opencl/CLConfig.py", line 134, in _autoSetNWorkUnits
    raise ValueError(f"The automatic test for optimal N_WORK_UNITS failed. Please manually run the test "
ValueError: The automatic test for optimal N_WORK_UNITS failed. Please manually run the test script 'testWorkUnits.py' and set N_WORK_UNITS in the config file at 'rayscattering/opencl/config.json'. 
... Error message: clBuildProgram failed: BUILD_PROGRAM_FAILURE - clBuildProgram failed: BUILD_PROGRAM_FAILURE - clBuildProgram failed: BUILD_PROGRAM_FAILURE

Build on <pyopencl.Device 'Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz' on 'Apple' at 0xffffffff>:

<program source>:61:6: warning: no previous prototype for function 'wangHash'
uint wangHash(uint seed){
     ^
<program source>:70:7: warning: no previous prototype for function 'getRandomFloatValue'
float getRandomFloatValue(__global unsigned int *seeds, unsigned int id){
      ^
<program source>:85:6: warning: no previous prototype for function 'normalizeVectorLocal'
void normalizeVectorLocal(float3 *vector){
     ^
<program source>:94:6: warning: no previous prototype for function 'normalizeVectorGlobal'
void normalizeVectorGlobal(__global float3 *vector){
     ^
<program source>:103:6: warning: no previous prototype for function 'rotateAroundAxisGlobal'
void rotateAroundAxisGlobal(__global float3 *mainVector, __global float3 *axisVector, float theta){
     ^
<program source>:128:6: warning: no previous prototype for function 'rotateAroundAxisLocal'
void rotateAroundAxisLocal(float3 *mainVector, float3 *axisVector, float theta){
     ^
<program source>:153:6: warning: no previous prototype for function 'rotateAround'
void rotateAround(float3 *mainVector, float3 *axisVector, float theta){
     ^
<program source>:178:8: warning: no previous prototype for function 'getAnyOrthogonalGlobal'
float3 getAnyOrthogonalGlobal(__global float3 *vector){
       ^
<program source>:185:8: warning: no previous prototype for function 'getAnyOrthogonal'
float3 getAnyOrthogonal(float3 *vector){
       ^
<program source>:210:7: warning: no previous prototype for function 'getScatteringDistance'
float getScatteringDistance(float mu_t, float randomNumber){
      ^
<program source>:214:7: warning: no previous prototype for function 'getScatteringAnglePhi'
float getScatteringAnglePhi(float randomNumber){
      ^
<program source>:219:7: warning: no previous prototype for function 'getScatteringAngleTheta'
float getScatteringAngleTheta(float g, float randomNumber){
      ^
<program source>:229:18: warning: no previous prototype for function 'getScatteringAngles'
ScatteringAngles getScatteringAngles(float rndPhi, float rndTheta,__global Photon *photons,
                 ^
<program source>:255:13: error: global variables must have a constant address space qualifier
const float EPS = 0.00001f;
            ^
<program source>:256:13: error: global variables must have a constant address space qualifier
const float EPS_CORRECTION = 0.0005f;
            ^
<program source>:257:13: error: global variables must have a constant address space qualifier
const float EPS_PARALLEL = 0.00001f;
            ^
<program source>:258:13: error: global variables must have a constant address space qualifier
const float EPS_SIDE = 0.000001f;
            ^
<program source>:300:21: warning: no previous prototype for function '_getBBoxIntersection'
GemsBoxIntersection _getBBoxIntersection(Ray ray, float3 minCornerVector, float3 maxCornerVector) {
                    ^
<program source>:370:6: warning: no previous prototype for function '_findBBoxIntersectingSolids'
void _findBBoxIntersectingSolids(Ray ray, Scene *scene, uint gid){
     ^
<program source>:387:6: warning: no previous prototype for function '_sortSolidCandidates'
void _sortSolidCandidates(Scene *scene, uint gid) {
     ^
<program source>:412:10: warning: no previous prototype for function '_getTriangleIntersection'
HitPoint _getTriangleIntersection(Ray ray, float3 v1, float3 v2, float3 v3) {
         ^
<program source>:460:14: warning: no previous prototype for function '_findClosestPolygonIntersection'
Intersection _findClosestPolygonIntersection(Ray ray, uint solidID,
             ^
<program source>:489:7: warning: no previous prototype for function '_cotangent'
float _cotangent(float3 v0, float3 v1, float3 v2) {
      ^
<program source>:495:6: warning: no previous prototype for function 'setSmoothNormal'
void setSmoothNormal(Intersection *intersection, __global Triangle *triangles, __global Vertex *vertices) {
     ^
<program source>:517:6: warning: no previous prototype for function '_composeIntersection'
void _composeIntersection(Intersection *intersection, Ray *ray, Scene *scene) {
     ^
<program source>:528:14: warning: no previous prototype for function 'findIntersection'
Intersection findIntersection(Ray ray, Scene *scene, uint gid) {
             ^
<program source>:586:7: warning: no previous prototype for function '_getReflectionCoefficient'
float _getReflectionCoefficient(float n1, float n2, float thetaIn) {
      ^
<program source>:613:6: warning: no previous prototype for function '_getIsReflected'
bool _getIsReflected(float nIn, float nOut, float thetaIn, __global uint *seeds, uint gid) {
     ^
<program source>:622:7: warning: no previous prototype for function '_getReflectionDeflection'
float _getReflectionDeflection(float thetaIn) {
      ^
<program source>:626:7: warning: no previous prototype for function '_getRefractionDeflection'
float _getRefractionDeflection(float nIn, float nOut, float thetaIn) {
      ^
<program source>:632:6: warning: no previous prototype for function '_createFresnelIntersection'
void _createFresnelIntersection(FresnelIntersection* fresnelIntersection,
     ^
<program source>:643:21: warning: no previous prototype for function 'computeFresnelIntersection'
FresnelIntersection computeFresnelIntersection(float3 rayDirection, Intersection *intersection,
                    ^
<program source>:677:11: error: global variables must have a constant address space qualifier
const int NO_SOLID_ID = -1;
          ^
<program source>:678:11: error: global variables must have a constant address space qualifier
const int NO_SURFACE_ID = -1;
          ^
<program source>:680:6: warning: no previous prototype for function 'moveBy'
void moveBy(float distance, __global Photon *photons, uint photonID){
     ^
<program source>:684:6: warning: no previous prototype for function 'scatterBy'
void scatterBy(float phi, float theta, __global Photon *photons, uint photonID){
     ^
<program source>:689:6: warning: no previous prototype for function 'decreaseWeightBy'
void decreaseWeightBy(float delta_weight, __global Photon *photons, uint photonID){
     ^
<program source>:693:6: warning: no previous prototype for function 'interact'
void interact(__global Photon *photons, __constant Material *materials, __global DataPoint *logger,
     ^
<program source>:705:6: warning: no previous prototype for function 'scatter'
void scatter(__global Photon *photons, __constant Material *materials, __global uint *seeds, __global DataPoint *logger,
     ^
<program source>:717:6: warning: no previous prototype for function 'roulette'
void roulette(float weightThreshold, __global Photon *photons, __global uint *seeds, uint gid, uint photonID){
     ^
<program source>:730:6: warning: no previous prototype for function 'reflect'
void reflect(FresnelIntersection *fresnelIntersection, __global Photon *photons, uint photonID){
     ^
<program source>:731:18: error: passing '__global __float3 *' to parameter of type '__float3 *' changes address space of pointer
    rotateAround(&photons[photonID].direction, &fresnelIntersection->incidencePlane, fresnelIntersection->angleDeflection);
                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
<program source>:153:27: note: passing argument to parameter 'mainVector' here
void rotateAround(float3 *mainVector, float3 *axisVector, float theta){
                          ^
<program source>:734:6: warning: no previous prototype for function 'refract'
void refract(FresnelIntersection *fresnelIntersection, __global Photon *photons, uint photonID){
     ^
<program source>:735:18: error: passing '__global __float3 *' to parameter of type '__float3 *' changes address space of pointer
    rotateAround(&photons[photonID].direction, &fresnelIntersection->incidencePlane, fresnelIntersection->angleDeflection);
                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
<program source>:153:27: note: passing argument to parameter 'mainVector' here
void rotateAround(float3 *mainVector, float3 *axisVector, float theta){
                          ^
<program source>:738:6: warning: no previous prototype for function 'logIntersection'
void logIntersection(Intersection *intersection, __global Photon *photons, __global Surface *surfaces,
     ^
<program source>:766:7: warning: no previous prototype for function 'reflectOrRefract'
float reflectOrRefract(Intersection *intersection, __global Photon *photons, __constant Material *materials,
      ^
<program source>:772:30: warning: comparison of integers of different signs: 'int' and '__global uint' (aka '__global unsigned int')
    if (solidIDTowardsNormal != photons[photonID].solidID) {
        ~~~~~~~~~~~~~~~~~~~~ ^  ~~~~~~~~~~~~~~~~~~~~~~~~~
<program source>:810:7: warning: no previous prototype for function 'propagateStep'
float propagateStep(float distance, __global Photon *photons, __constant Material *materials, Scene *scene,
      ^
<program source>:838:38: warning: comparison of integers of different signs: 'int' and '__global uint' (aka '__global unsigned int')
            if (solidIDTowardsNormal != photons[photonID].solidID) {
                ~~~~~~~~~~~~~~~~~~~~ ^  ~~~~~~~~~~~~~~~~~~~~~~~~~

(options: -I /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/pyopencl/cl)
(source saved as /var/folders/vw/xyjt8vln3yb962l7lzwbgvrc0000gn/T/tmpj4dgkn0z.cl)

```

Reply:
Your version of OpenCL C seems to be more restrictive than ours. I attempted a fix for the two errors in your traceback. If there are no remaining errors, the code will run and the warnings should not be displayed. Still, we could also tackle these warnings so that future error tracebacks are easier to read and debug.